libnd4j/include/ops/declarable/helpers/cpu/transforms_merge.cpp

```diff
@@ -0,0 +1,269 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+// @author Yurii Shyrma (iuriish@yahoo.com), created on 20.04.2018
+// @author Oleh Semeniv (oleg.semeniv@gmail.com)
+//
+
+#include <ops/declarable/helpers/transforms.h>
+#include <helpers/Loops.h>
+
+namespace sd 	  {
+namespace ops 	  {
+namespace helpers {
+
+
+//////////////////////////////////////////////////////////////////////////
+template<typename T>
+static void mergeMaxIndex_(const std::vector<NDArray*>& inArrs, NDArray& output) {
+
+    const Nd4jLong numArgs = inArrs.size();
+    auto x = inArrs[0];
+
+    auto func = PRAGMA_THREADS_FOR {
+        for (auto e = start; e < stop; e++) {
+            T max = -DataTypeUtils::max<T>();
+            Nd4jLong idx = 0;
+
+            for (Nd4jLong i = 0; i < numArgs; i++) {
+                T v = inArrs[i]->e<T>(e);
+                if (v > max) {
+                    max = v;
+                    idx = i;
+                }
+            }
+            output.p(e, idx);
+        }
+    };
+
+    samediff::Threads::parallel_for(func, 0, x->lengthOf());
+}
+
+void mergeMaxIndex(sd::LaunchContext * context, const std::vector<NDArray*>& inArrs, NDArray& output) {
+    BUILD_SINGLE_SELECTOR(inArrs[0]->dataType(), mergeMaxIndex_, (inArrs, output), LIBND4J_TYPES);
+}
+
+
+//////////////////////////////////////////////////////////////////////////
+template<typename T>
+static void mergeMax_(const std::vector<NDArray*>& inArrs, NDArray& output) {
+    const Nd4jLong numArgs = inArrs.size();
+    auto x = inArrs[0];
+
+    auto func = PRAGMA_THREADS_FOR {
+        for (auto e = start; e < stop; e++) {
+            T max = -DataTypeUtils::max<T>();
+            for (Nd4jLong i = 0; i < numArgs; i++) {
+                T v = inArrs[i]->e<T>(e);
+                if (v > max)
+                    max = v;
+            }
+            output.p(e, max);
+        }
+    };
+
+    samediff::Threads::parallel_for(func, 0, x->lengthOf());
+}
+
+void mergeMax(sd::LaunchContext * context, const std::vector<NDArray*>& inArrs, NDArray& output) {
+    BUILD_SINGLE_SELECTOR(output.dataType(), mergeMax_, (inArrs, output), LIBND4J_TYPES);
+}
+
+
+//////////////////////////////////////////////////////////////////////////
+template<typename T>
+static void mergeMaxBp_(const std::vector<NDArray*>& inArrs, std::vector<NDArray*>& outArrs) {
+    
+    const Nd4jLong numArgs = outArrs.size();
+    // last array is gradient
+    auto gradient = inArrs[numArgs]->bufferAsT<T>();
+    auto length = inArrs[numArgs]->lengthOf();
+    
+    bool bSameOrderAndEws1 = true;
+
+    auto ordering = inArrs[numArgs]->ordering();
+    for (auto v : inArrs) {
```

Comment:
let check here the possibility to use same offsets in slower loop, I mean
... &= shape::haveSameShapeAndStrides

Reply:
corrections added per request. Files names changed.